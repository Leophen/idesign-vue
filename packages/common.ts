import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
