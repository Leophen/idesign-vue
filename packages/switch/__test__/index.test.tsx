import { vi } from 'vitest';
import { Switch } from '../index';
import { render } from '../../common'

describe('Switch 组件测试', () => {
  it('create', () => {
    const wrapper = render(<Switch />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('value', () => {
    const wrapper = render(<Switch modelValue={true} />);
    expect(wrapper.classes()).toContain('i-switch-is-checked');
  });

  it('defaultValue', () => {
    const wrapper = render(<Switch defaultValue={true} />);
    expect(wrapper.classes()).toContain('i-switch-is-checked');
  });

  it('onChange', () => {
    const changeFn = vi.fn();
    const wrapper = render(<Switch onChange={changeFn} />);
    wrapper.find('.i-switch').trigger('click');
    expect(changeFn).toHaveBeenCalled();
  });

  it('inactiveColor', () => {
    const wrapper = render(<Switch inactiveColor="blue" />);
    expect(wrapper.element.getAttribute('style')).toMatch('background-color: blue');
  });

  it('activeColor', () => {
    const wrapper = render(<Switch activeColor="red" modelValue={true} />);
    expect(wrapper.element.getAttribute('style')).toMatch('background-color: red');
  });

  it('size', () => {
    const wrapper = render(<Switch size="small" />);
    expect(wrapper.classes()).toContain('i-switch--size-small');
  });

  it('inactiveLabel', () => {
    const wrapper = render(
      <Switch
        v-slots={{
          inactiveLabel: () => '关'
        }}
      />
    );
    expect(wrapper.find('.i-switch__content').text()).toBe('关');
  });

  it('activeLabel', () => {
    const wrapper = render(
      <Switch
        modelValue={true}
        v-slots={{
          activeLabel: () => '开'
        }}
      />
    );
    expect(wrapper.find('.i-switch__content').text()).toBe('开');
  });

  it('disabled', () => {
    const wrapper = render(<Switch disabled />);
    expect(wrapper.classes()).toContain('i-switch-is-disabled');
  });
});

