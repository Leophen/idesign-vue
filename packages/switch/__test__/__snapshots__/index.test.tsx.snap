// Vitest Snapshot v1

exports[`Switch ç»„ä»¶æµ‹è¯• > create 1`] = `
<button
  class="i-switch"
  role="switch"
  type="button"
>
  <span
    class="i-switch__handle"
  >
    <!--v-if-->
  </span>
  <div
    class="i-switch__content"
  >
    
    
    <!--v-if-->
  </div>
</button>
`;
